src/message_bus.zig

```diff
@@ -73,60 +191,413 @@ pub const MessageBus = struct {
-            // synchronous append in on_message() is executed the next iteration of the outer loop.
-            // This is not critical for safety here, but see run() in src/io.zig where it is.
-            var copy = self.ready;
-            self.ready = .{};
-            while (copy.pop()) |envelope| {
-                assert(envelope.message.references > 0);
-                switch (envelope.address) {
-                    .replica => |r| r.on_message(envelope.message),
-                }
-                envelope.message.references -= 1;
-                self.gc(envelope.message);
-                // The lifetime of an envelope will often be shorter than that of the message.
-                self.allocator.destroy(envelope);
+    pub fn send_header_to_client(self: *MessageBus, client_id: u128, header: Header) void {
+        assert(header.size == @sizeOf(Header));
+
+        // TODO Pre-allocate messages at startup.
+        var message = self.create_message(@sizeOf(Header)) catch unreachable;
+        message.header.* = header;
+
+        const data = message.buffer[@sizeOf(Header)..message.header.size];
+        // The order matters here because checksum depends on checksum_data:
+        message.header.set_checksum_data(data);
+        message.header.set_checksum();
+
+        assert(message.references == 0);
+        self.send_message_to_client(replica, message);
+    }
+
+    /// Try to send the message to the client with the given id.
+    /// If the client is not currently connected, the message is silently dropped.
+    pub fn send_message_to_client(self: *MessageBus, client_id: u128, message: *Message) void {
+        for (self.connections) |connection| {
+            switch (connection) {
+                .client => |id| if (id == client_id) {
+                    message.references += 1;
+                    // TODO Pre-allocate envelopes at startup.
+                    const envelope = self.allocator.create(Envelope) catch unreachable;
+                    envelope.* = .{
+                        .address = self.configuration[replica],
+                        .message = message,
+                    };
+                    connection.send_message(envelope);
+                    return;
+                },
+                else => {},
             }
         }
+        // TODO: not sure if this is necessary or not
+        self.message_bus.gc(message);
+    }
+
+    pub fn flush(self: *MessageBus) void {
+        // Deliver messages the server replica has sent to itself.
+        // Iterate on a copy to avoid a potential infinite loop.
+        var copy = self.replicas[self.server.replica].send_queue;
+        self.replicas[self.server.replica].send_queue = .{};
+        while (copy.pop()) |envelope| {
+            self.server.on_message(envelope.message);
+            envelope.message.references -= 1;
+            self.message_bus.gc(envelope.message);
+            self.message_bus.allocator.destroy(envelope);
```

Comment:
`self.message_bus` should be `self` since we are already within the context of the message bus.